Blaze: Add picker view for target languages

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11630 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds a new screen for selecting target languages for a creating a new Blaze campaign. Changes include:
- Updated all Blaze target option models to conform to `Hashable` and `Identifiable`. This is necessary to display the models on a SwiftUI list.
- Renamed `SelectionList` to `SingleSelectionList`. For simplicity, I created a separate reusable view `MultiSelectionList` for use in the Blaze campaign creation flow.
- Created `BlazeTargetLanguagePickerView` and view model to sync and display languages for picking.
- Integrated the language picker view to the Blaze campaign creation form.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Log in to a store with at least one existing product and is eligible for Blaze.
- Navigate to the Products tab and select any public product.
- On the product form, select Promote with Blaze.
- Notice that the campaign creation form is presented. Select the language field.
- Notice that a new screen is displayed for selecting target languages. Confirm that the selection and filtering work as expected.
- Select Cancel, the view should be dismissed without any changes to the campaign creation form.
- Select the language field again, this time select a different option and tap Save. The language field in the creation form should be updated correctly.


## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/5533851/a4a12605-5e74-4c03-a0a5-fc3d4876d608



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Rename SelectionList to SingleSelectionList

itsmeichigo committed:
Add a new reusable type for multiple selection list

itsmeichigo committed:
Rename MultipleSelectionList to MultiSelectionList

Because it sounds cooler?

itsmeichigo committed:
Make Blaze target option models to conform to Identifiable and Hashable

itsmeichigo committed:
Add view model for BlazeTargetLanguagePickerView

itsmeichigo committed:
Add localized strings for target language picker

itsmeichigo committed:
Update selection behavior for MultiSelectionList

itsmeichigo committed:
Sync target languages

itsmeichigo committed:
Add target language picker to campaign creation form

itsmeichigo committed:
Update text for target languages

itsmeichigo committed:
Handle toggling selection for multi selection list

itsmeichigo committed:
Sort language list

itsmeichigo committed:
Rename group for target languages

itsmeichigo committed:
Add unit tests for BlazeTargetLanguagePickerViewModel

itsmeichigo committed:
Update sort logic

itsmeichigo committed:
Fix typo in comment

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift:
```diff
@@ -22,16 +22,19 @@ final class BlazeCampaignCreationFormViewModel: ObservableObject {
             return nil
         }
-        return BlazeTargetOptions(locations: locations, languages: languages, devices: devices, pageTopics: pageTopics)
+        return BlazeTargetOptions(locations: locations?.map { $0.id },
+                                  languages: languages?.map { $0.id },
+                                  devices: devices?.map { $0.id },
+                                  pageTopics: pageTopics?.map { $0.id })
```
nit: Consider unwrapping these values in the above `guard` statement. 

```swift
        guard let locations, let languages, let devices, let pageTopics else {
            return nil
        }
        return BlazeTargetOptions(locations: locations.map { $0.id },
                                  languages: languages.map { $0.id },
                                  devices: devices.map { $0.id },
                                  pageTopics: pageTopics.map { $0.id })
```

itsmeichigo replied:
This is not correct. We don't expect every option to be available to create a non-nil option object. I'll keep my change as-is.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/SingleSelectionList.swift:
```diff
@@ -77,7 +78,7 @@ struct SelectionList<T: Hashable>: View {
 struct SelectionList_Previews: PreviewProvider {
     static var previews: some View {
-        SelectionList(title: "Lunch",
+        SingleSelectionList(title: "Lunch",
```
Super nit: Indentation needs to be updated.

itsmeichigo replied:
Updated in 9137b6fd0a32b44501ee83f05bdbeeef06dec8fa

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Blaze/BlazeTargetLanguagePickerViewModelTests.swift:
```diff
@@ -0,0 +1,90 @@
+        }
+    }
+
+    func test_languages_filters_matching_languages_if_searchQuery_is_empty() {
```
```suggestion
    func test_languages_filters_matching_languages_if_searchQuery_is_not_empty() {
```

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsTimeRangeCard.swift:
```diff
@@ -37,7 +37,7 @@ struct AnalyticsTimeRangeCard: View {
         createTimeRangeContent()
             .sheet(isPresented: $showTimeRangeSelectionView) {
-                SelectionList(title: Localization.timeRangeSelectionTitle,
+                SingleSelectionList(title: Localization.timeRangeSelectionTitle,
```
Super nit: Indentation fix needed. This applies to other places where we have replaced `SelectionList` with `SingleSelectionList`.

itsmeichigo replied:
Updated in 9137b6fd0a32b44501ee83f05bdbeeef06dec8fa

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/TargetLanguages/BlazeTargetLanguagePickerViewModel.swift:
```diff
@@ -0,0 +1,97 @@
+final class BlazeTargetLanguagePickerViewModel: ObservableObject {
+
+    @Published var searchQuery: String = ""
+    @Published private(set) var languages: [BlazeTargetLanguage] = []
```
nit: Consider adding a comment to explain that holds the `languages` after filtering based on search query.

itsmeichigo replied:
Added in 2533ea0ca946a5582f30cb0d89dcca28e8d3284f.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/MultiSelectionList.swift:
```diff
@@ -0,0 +1,134 @@
+                        }
+                        .contentShape(Rectangle())
+                        .onTapGesture {
+                            if selectedItems.isEmpty {
+                                selectedItems = Set(contents)
+                            } else {
+                                selectedItems = []
+                            }
```
![AllSelection](https://github.com/woocommerce/woocommerce-ios/assets/524475/ba5ebbdf-0cd4-43a5-9b6f-55fb5dd89e4d)

Should we select all items when `selectedItems.isEmpty` is empty? 

---

https://github.com/woocommerce/woocommerce-ios/assets/524475/5b146fb3-9097-4d71-a04c-7a2e335244cd


In the above scenario, I have selected all the available languages, but I don't see "All" being displayed in the creation form. 

I am not sure what is the expected behaviour here.

itsmeichigo replied:
Good point! I updated the logic in be41f3e8ba21826521651a120994e41a8c830773 - the All row should be selected if all other rows are and not when any of the other rows are unselected.

selanthiraiyan reviewed (approved):
Thanks for working on adding a `MultiSelectionList`, Huong! 

This looks great. I left a few non-blocking nits. 🚀

itsmeichigo committed:
Update WooCommerce/WooCommerceTests/ViewRelated/Blaze/BlazeTargetLanguagePickerViewModelTests.swift

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

itsmeichigo committed:
Fix indentation for SingleSelectionList

itsmeichigo committed:
Add comment for `languages`

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Update selection logic for MultiSelectionList

itsmeichigo committed:
Merge branch 'trunk' into feat/11630-language-picker

itsmeichigo commented:
Thank you @selanthiraiyan for the reviews! I addressed the issues mentioned in your comments - will merge now to continue with other tasks. Let me know if you have any other comments, I'll handle them in the next PRs.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
